src/GafferSceneUI/LightPositionTool.cpp

```diff
@@ -0,0 +1,347 @@
+		( m_worldCentroid - shadowPivot ).length() +
+		shadowPivot
+	;
+
+	const M44f centroidTransform = rotationMatrix( m_worldCentroid - shadowPivot, newCentroid - shadowPivot );
+
+	UndoScope undoScope( selection().back().editTarget()->ancestor<ScriptNode>() );
+
+	for( const auto &s : selection() )
+	{
+		// See `RotateTool::buttonPress()` for a description of why we use this relatively
+		// elaborate orientation calculation.
+		Context::Scope scopedContext( s.context() );
+
+		ScenePlug::ScenePath parentPath( s.path() );
+		parentPath.pop_back();
+
+		const M44f worldParentTransform = s.scene()->fullTransform( parentPath );
+		const M44f worldParentTransformInverse = worldParentTransform.inverse();
+		const M44f localTransform = s.scene()->transform( s.path() );
+
+		const M44f worldTransform = s.scene()->fullTransform( s.path() );
+
+		const V3f offset = (
+			( ( worldTransform.translation() - shadowPivot ) * centroidTransform ) + shadowPivot -
+			worldParentTransform.translation() - localTransform.translation()
+		);
+
+		V3f currentYAxis;
+		localTransform.multDirMatrix( V3f( 0.f, 1.f, 0.f ), currentYAxis );
+
+		// Point in the pivot-shadowPoint direction, in local space
+		const V3f targetZAxis = ( shadowPoint - shadowPivot ) * worldParentTransformInverse;
+
+		M44f orientationMatrix = rotationMatrixWithUpDir(
+			V3f( 0.f, 0.f, -1.f ), targetZAxis, currentYAxis
+		);
+
+		V3f originalRotation;
+		extractEulerXYZ( localTransform, originalRotation );
+		const M44f originalRotationMatrix = M44f().rotate( originalRotation );
+
+		const M44f relativeMatrix = originalRotationMatrix.inverse() * orientationMatrix;
+
+		V3f relativeRotation;
+		extractEulerXYZ( relativeMatrix, relativeRotation );
+
+		TranslationRotation( s ).apply( offset, relativeRotation );
+	}
+}
+
+bool LightPositionTool::affectsHandles( const Gaffer::Plug *input ) const
+{
+	if( TransformTool::affectsHandles( input ) )
+	{
+		return true;
+	}
+
+	return input == scenePlug()->transformPlug();
+}
+
+void LightPositionTool::updateHandles( float rasterScale )
+{
+
+}
+
+void LightPositionTool::selectionChanged( const TransformTool &tool )
+{
+	if( activePlug()->getValue() )
+	{
+		if( selection().empty() )
+		{
+			return;
+		}
+
+		m_worldCentroid = V3f( 0 ) * selection().back().orientedTransform( Orientation::World );
+	}
+}
+
+void LightPositionTool::plugSet( Plug *plug )
+{
+	if( plug == activePlug() && !activePlug()->getValue() )
+	{
+		m_shadowPivot = std::nullopt;
+		m_shadowPoint = std::nullopt;
+	}
+}
+
+void LightPositionTool::plugInputChanged( Plug *plug )
+{
+	if( activePlug()->getValue() && plug == view()->inPlug() )
+	{
+		m_shadowPivot = std::nullopt;
+		m_shadowPoint = std::nullopt;
+	}
+}
+
+bool LightPositionTool::buttonPress( const ButtonEvent &event )
+{
+	if(
+		event.button != ButtonEvent::Left ||
+		!activePlug()->getValue() ||
+		!(
+			event.modifiers == ButtonEvent::Control ||
+			event.modifiers == ButtonEvent::Shift
+		)
+	)
+	{
+		return false;
```

Comment:
When testing, I got into a few unfortunate situations where I ended up positioning something that wasn't a light. I think part of the problem was this `return false`, which would let me accidentally select some geometry with a left click. Once I'd done that, my immediate instinct was to `Ctrl + Click` to deselect the bit of geometry, but of course that's setting a pivot, which moves the selected object.

I'm not sure what the solution is, but I wonder if we need to either expose _all_ the usual selection behaviours, or _none_ of them? If the former, then we'd need different keypresses, in which case perhaps `V` could work for the "shadow point" click, somewhat matching the snapping behaviours in the TranslateTool and RotateTool?

Reply:
I'm not sure if it's related to this or not, but I've also been seeing a bunch of the following :

```WARNING : TransformTool::updateSelection : Last selected path not included in selection```

Here's a simple repro :

1. Open Gaffer.
2. Make a Cube node and view its output.
3. Enable the LightPositionTool.
4. Left-click the cube in the Viewer.

Reply:
Got to the bottom of this one - it's a result of what I was doing in my `selectionChangedSignal()` handler :
1. `SelectionTool` handles a click event and calls `ContextAlgo::setSelectedPaths()`, _before_ `ContextAlgo::setLastSelectedPath()` https://github.com/GafferHQ/gaffer/blob/2d22be25ee4aabe54f85cb9a8dce4e7bf1f4dd7f/src/GafferSceneUI/SelectionTool.cpp#L241
2. Setting the selected paths fires my `selectionChangedSignal()` handler
3. I was calling `TransformTool::selection()`, which does `updateSelection()`
4. `updateSelection()` assumes we have a proper `lastSelectedPath` and gives the warning noted above if there's nothing there https://github.com/GafferHQ/gaffer/blob/2d22be25ee4aabe54f85cb9a8dce4e7bf1f4dd7f/src/GafferSceneUI/TransformTool.cpp#L1180-L1192

I think this will all be solved when I add an actual handle, so I'll move my centroid code to `updateHandles()` where I think it is maybe more appropriate.

I don't know if there's anything to be improved with the `selectionChangedSignal()` behavior - it seems logical that in that handler I could ask for the current selection? But that's not part of this effort.

Reply:
Good sleuthing!

> I don't know if there's anything to be improved with the selectionChangedSignal() behavior - it seems logical that in that handler I could ask for the current selection?

Yep, completely. The current behaviour is breaking one of the fundamental rules of signals - don't emit them while your visible state is inconsistent.

> I'll move my centroid code to updateHandles() where I think it is maybe more appropriate.

Yep, that makes sense (although I think we're dropping the centroid anyway aren't we?). I suspect the reason this hasn't become a problem in the other tools is because they are following the "don't do any work unless visible" rule, so everything is deferred until `preRender()`, at which point the selection state is consistent.